### DISCUSSION TITLE
Internal server error

### DISCUSSION THREAD

**Discussion by mariuszmalek** [brefphp/bref#953] [Support]
Hello, thanks for the answers to my previous problems. I was learning about serverless and I fixed everything I wrote about by setting up NAT Gateway.

Now I have another problem. During registration (user normal register but not login after) in the application, I have this error: ** "Internal server error" **. I found where the bug is. I set the listener (Laravel) to check the user's parameters, and when everything is fine, sending him a welcome e-mail. When I hide that part code of a sending email, I log in as usual.

Do you know why I have that problem?

**Comment by mnapoli** [comment]
"Internal server error" means there is an error in your application. You need to look at the logs and investigate yourself: https://bref.sh/docs/environment/logs.html

This is a generic error message, nobody can help you with just that message.

**Reply by mariuszmalek** [reply]
Problem is that in logs, I don't see any Laravel errors.

**Reply by mariuszmalek** [reply]
Example: 

![image](https://user-images.githubusercontent.com/26696860/121770379-484de080-cb69-11eb-9554-9cea9f4e7ff5.png)

**Reply by t-richard** [reply]
Timeouts are generally a sign of a connection timing out in the code (http connection, database connection, etc).

Are you sure the Nat gateway is correctly set up in all your lambda?

It looks like the email sending is timing out because it's unable to connect to the email provider you are using.

If a Nat gateway is correctly set up, it might be your outbound security group that is not allowing the specific port of the protocol your are using (80 for http,443 for https, 25 for smtp, etc)

**Reply by mariuszmalek** [reply]
Thank you very much for such a quick answer. You are wonderful. I will try to fix it with your tips.

**Reply by mnapoli** [reply]
https://bref.sh/docs/environment/database.html#accessing-the-internet

**Reply by mariuszmalek** [reply]
I do as it is in the documentation but still have a problem. This is what it looks like. Only once a day does it work well, but after generating one email, the software always returns an error.

Is it possible the problem is elsewhere? Now I also noticed that sending photos doesn't even work for me, but here I already have another message. I get the error "s3 Error executing" PutObject ", and I set the permissions in IAM and the serverless file correctly.

**Reply by mariuszmalek** [reply]
I found the problem. When software returns one error then Lambda can of blocked? (I don't understand). I fix this by change Timeout (from 28s to 1minute) in Lambda anytime when the software block in Internal Server Error.

**Reply by t-richard** [reply]
This first Lambda invocation starts from scratch, AWS will load your code, layers then Bref will initialize, laucnh FPM, etc. This is generally referred as cold start. AWS keeps this now "warm" Lambda aside for future use for about 10 minutes. 

This way, the next invocations will be faster because there is already a Lambda that is warmed (with your code deployed in it, with PHP-FPM started, etc)

If for some reason, the first invocation messed up with the Lambda, then next invocations could error. When changing the timeout, mémory or similar things in Lambda, AWS will invalidate all the "warm" Lambdas so the following requests will start from scratch. This is also the case when you deploy a new version.

So by changing the timeout, I don't think you are solving the problem, you are just resetting the system until the error occurs again. If I'm right, the error would go away if you retry 10-15 minutes after that.

**Reply by mariuszmalek** [reply]
It is true. After the reset, it all returns to normal, but until a new error occurs. I can't find a solution anymore.

**Reply by mariuszmalek** [reply]
@t-richard, is possible to fix that? I mean to when some kind of an error from software return then Lambda always works well?

**Reply by t-richard** [reply]
This is hard to troubleshoot. AFAIK there is no way to tell Lambda to discard a warm instance (and you don't what that anyway).

The only way is to find the real issue and make sure it never happens. Easier said than done, I agree.

It's really hard to be able help you on that without a clear error message, a stack trace, etc. 

A simple PHP exception should not put your Lambda in a such state IMO.

**Reply by mariuszmalek** [reply]
@t-richard, now I don't understand anything. Software not working and in logs I have "2021-06-16T12:44:31.311Z 6a91a0de-2c2a-4475-a9d7-0007eff1dff1 Task timed out after 28.03 seconds". Error is only when I use Queue. I can normally log in to the software.

I do everything that is in the documentation. I opened internet access by NAT Gateway.

**Reply by mariuszmalek** [reply]
Still not working.. Nothing help me.
